azure/ConsiderationsForServiceDesign.md

```diff
@@ -229,6 +229,33 @@ An important distinction between RELO and status monitor LROs is that there is a
+- Idempotent APIs solve a whole class of network issues where customers have no idea how to proceed if they send a request to the service but never get a response.
+- Accessing resources from multiple microservices can quickly lead to complex race conditions so conditional requests provide optimistic concurrency for safe usage.
+- Reframing the purpose of an API can obviate some errors.  This is most often specific to your operations, but an example from the Guidelines is thinking about `DELETE`s as _"ensure no resource at this location exists"_ so they can return an easier to use `204` instead of _"delete this exact resource instance"_ which would fail with a non-idempotent `404`.
+
+There are two types of errors returned from your service and customers handle them differently.
+- Usage errors where the customer is calling your API incorrectly.  The customer can easily make these errors go away by fixing their code.  We expect most usage errors to be found during testing.
+- Runtime errors that can't be prevented by the customer and need to be recovered from.  Some runtime errors like `429` throttling will be handled automatically by client libraries, but most will be situations like a `409` conflict requiring knowledge about the customer's application to remedy.
+
+We provide appropriate [HTTP status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses) for customers to handle errors generically and error code strings in our common error schema and the `x-ms-error-code` header for customers to handle errors specifically.  As an example, consider what a customer would do when trying to get the properties of a Storage blob:
+- A `404` status code tells them the blob doesn't exist and the customer can report the error to their users
+- A `BlobNotFound` or `ContainerNotFound` error code will tell them why the blob doesn't exist so they can take steps to recreate it
+
+The common error schema in the Guidelines allows nested details and inner errors that have their own error codes, but the top-level error code is the most important.  The HTTP status code and the top-level error code are the only part of your error that we consider part of your API contract that follows the same compatibility requirements as the rest of your API.  Importantly, this means you **can not change the HTTP status code or top-level error code for an API in GA'ed service version**.  You can only return new status codes and error codes in future API versions if customers make use of new features that trigger new classes of errors.  Battle tested error handling is some of the hardest code to get right and we can't break that for customers when they upgrade to the latest version.  The rest of the properties in your error like `message`, `details`, etc., are not considered part of your API contract and can change to improve the diagnosability of your service.
```

Comment:
And the top-level exception type, right?

Reply:
When we say the message is not part of the API contract and can change, does it mean we can change the messages to a GAed service to improve diagnosability?

Reply:
Yes, for any given operation, the only parts to the contract are the HTTP status code & the error code string - customers can write code that accesses these values. ALL other values are not part of the contract and may be changed by the service as desired. In our Go SDK, our error type only has fields for the HTTP status code & error code string and then raw response - we do no expose the other fields because we don't want customers to write code that access them. They can get the values from the raw response which will give them a JSON string which they have to parse to do whatever they want to do with it.

Reply:
The top-level exception type is going to be a language by language thing and that'll version according to the language guidelines.  C#, for example, has lots of specific rules in section D.12 at https://aka.ms/fxdg3/breaking (visit https://aka.ms/oreilly once for SSO).

Yes, it's possible to change error messages in GA'ed services to make them more diagnosable.  I'd probably recommend waiting until the next version bump if possible though because _someone_ is still probably parsing it even though they shouldn't.

Reply:
I understand that we can change the error message in the next API version.  But to improve customer satisfaction, can we make hotfixes to improve the error message for previously released GA versions?  Specifically, can we add a hyperlink to the error message to provide additional information?

Reply:
@JeffreyRichter should answer definitively, but I think it's a reasonable to update an error message without a version bump if you think there's significant customer value.  I'd still try to append to rather than edit the existing error message if possible to minimize the potential impact when we're changing without a version bump.

Reply:
Yes, as error messages are not part of the contract, they can be updated without a version bump. Just like improving the performance of an operation can be done without a version bump since perf is not part of the contract either.